spec/Section 3 -- Type System.md

```diff
@@ -1340,14 +1359,283 @@ Union type extensions have the potential to be invalid if incorrectly defined.
+
+Tagged types represent a list of possible named member fields, exactly one of
+which must be present. This resulting member field must yield a value of a
+specific type.  In outputs, like for Objects, Tagged types should be serialized
+as ordered maps: the queried member field names (or aliases) that are present
+are the keys and the result of evaluating the member field is the value,
+ordered by the order in which they appear in the query. Note that Tagged types
+may be objects with more than one key even though they guarantee that exactly
+one member field is present, this is because aliases may be used, and
+`__typename` may also be queried (and will return the name of the tagged type).
+
+All member fields defined within a Tagged type must not have a name which
+begins with {"__"} (two underscores), as this is used exclusively by GraphQL's
+introspection system.
+
+For example, a type `PackQuantity` could be described as:
+
+```graphql example
+tagged PackQuantity {
+  volumeInMetresCubed: Float!
+  numberOfBoxes: Int!
+  description: String!
+}
+```
+
+In this case we're representing exactly one of the following:
+
+- an object containing a single key `volumeInMetresCubed` with an {Float} value
+- an object containing a single key `numberOfBoxes` with a {Int} value
+- an object containing a single key `description` with an {String} value
+
+The `PackQuantity` Tagged type is valid as both an input and output type, but
+this is not true of all tagged types. If a Tagged type has a member field whose
+type is only valid for output, then the Tagged type is only valid for output.
+If a Tagged type has a member field whose type is only valid for input, then
+the Tagged type is only valid for input. If a Tagged type has a member field
+that is only valid for input, and another member field that's only valid for
+output, then it is not a valid Tagged type.
+
+A member field of a Tagged type may be of any valid GraphQL type, including
+Scalar, Enum, Input Object type, Object type, an Interface, a Union, another
+Tagged type, or any wrapping type (e.g. list, non-null, or any combination
+thereof) whose underlying base type is one of those seven.
+
+Selecting all the member fields of our `PackQuantity` type:
+
+```graphql example
+{
+  volumeInMetresCubed
+  numberOfBoxes
+  description
+}
+```
+
+Could yield one of the following results:
+
+- `{ "volumeInMetresCubed": 3.8 }`
+- `{ "numberOfBoxes": 27 }`
+- `{ "description": "4 large boxes, 7 medium boxes, and 16 small boxes" }`
+
+Valid queries must supply a nested field set for a field that returns a Tagged
+type, so for this schema:
+
+```graphql example
+type Query {
+  packQuantity: PackQuantity
+}
+```
+
+This query is not valid:
+
+```graphql counter-example
+{
+  packQuantity
+}
+```
+
+However, this query is valid:
+
+```graphql example
+{
+  packQuantity {
+    numberOfBoxes
+  }
+}
+```
+
+And may yield one of the following results:
+
+- `{ "numberOfBoxes": 27 }`
+- `{}`
```

Comment:
Permitting empty return objects is one of the parts of the output side of tagged types that makes me uncomfortable - although I can't articulate the exact rule, it feels in violation of some GraphQL principle somewhere. It's also somewhat inconsistent with how it's used on input, where providing 0 fields to a tagged type is invalid.

Because of this I am leaning more strongly towards making this an input-only type to start with.

Reply:
We already permit empty return objects:

```graphql
{
  __typename @skip(if: true)
}
```

Reply:
(We also allow it with fragment spreads; e.g. [this SWAPI example](https://graphql.org/swapi-graphql/?query=%7B%0A%20%20node(id%3A%22cGVvcGxlOjI%3D%22)%20%7B%0A%20%20%20%20...%20on%20Species%20%7B%0A%20%20%20%20%20%20name%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D)

```graphql
{
  node(id:"cGVvcGxlOjI=") {
    ... on Species {
      name
    }
  }
}
```
)